File: 34_PathInTree/PathInTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def reConstructBinaryTree(pre, lin):
    if not pre and not lin:
        return None
    root = TreeNode(pre[0])
    if set(pre) != set(lin):
        return None
    i = lin.index(pre[0])
    root.left = reConstructBinaryTree(pre[1:i + 1], lin[:i])
    root.right = reConstructBinaryTree(pre[i + 1:], lin[i + 1:])
    return root


class Solution:
    # 返回二维列表，内部每个列表表示找到的路径
    def FindPath(self, root, expectNumber):
        # write code here
        if not root:
            return []
        if root.val == expectNumber and not root.left and not root.right:
            return [[root.val]]
        result = []
        left = self.FindPath(root.left, expectNumber - root.val)
        right = self.FindPath(root.right, expectNumber - root.val)
        for i in left + right:
            result.append([root.val] + i)
        return result


t = reConstructBinaryTree([10, 5, 4, 7, 12], [4, 5, 7, 10, 12])
s = Solution()
print(s.FindPath(t, 22))

File: 34_PathInTree/PathInTree.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <cstring>
using namespace std;

struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
  public:
    vector<vector<int>> buffer;
    vector<int> tmp;
    vector<vector<int>> FindPath(TreeNode *root, int expectNumber)
    {
        if (root == nullptr)
            return buffer;
        tmp.push_back(root->val);
        if (expectNumber == root->val && root->left == nullptr && root->right == nullptr)
            buffer.push_back(tmp);
        FindPath(root->left, expectNumber - root->val);
        FindPath(root->right, expectNumber - root->val);
        if (tmp.size())
            tmp.pop_back();
        return buffer;
    }
};